documentation/architecture/messaging/Reliability.md

```diff
@@ -0,0 +1,201 @@
+
+Exactly once delivery is controversal and hard to achieve in presence of errors.
+For practical purposes we're going to use more relaxed version of this definition
+where `DM = {R -> 1, UR -> 1}`
+
+<img src="./images/delivery_modes.jpg" width="100%">
+
+
+## Pipeline reliability
+
+When pipelining deliveries, e.g. `A->B ; B->C`, reliability and uniqness rates are multiplying, because they are probabilities of dependent events
+
+<img src="./images/pipeline_delivery.jpg" width="100%">
+
+Which means that 
+`R(A->B ; B->C) = R(A->B) * R(B->C)`
+and `UR(A->B ; B->C) = UR(A->B) * UR(B->C)`
+
+This means that if we pipeline deliveries:
+
+- at-most-once with at-least-once - the pipeline will be unreliable
+- exactly-once with at-least-once - the pipeline will be at-least-once
+- exactly-once with at-most-once - the pipeline will be at-most-once
+- unreliable to anything - the pipeline will be unreliable
+
+As you can see preserving high reliability in pipelined delivery is problematic, therefore reliability is better addressed by end-to-end wrapping
+
+## Local delivery
+
+**Local delivery is supposed to be implemented as highly reliable**
+
+Workers supposed to have delivery mode close to exactly once **as long as both workers are alive and don't have errors**
+
+For practical purposes delivery mode over a route with single local address `DM([0#B])` is assumed `{R->1, UR->1}`
+
+<img src="./images/local_delivery.jpg" width="100%">
+
+## Pipe reliability
+
+Since local delivery mode is close to exactly-once, we can assume that reliability of a delivery via pipe `P1->P2`, which connects to endpoints `A` and `B` on local routes `A->0#P1 ; P1->P2 ; P2->0#B` can be described as number of unique messages sent by `P2` divided by number of messages received by `P1`.
+
+<img src="./images/delivery_pipe.jpg" width="100%">
+
+We call that **pipe reliability**
+
+Same for uniqness rate we can define **pipe uniqness rate**
+
+This allows us to design the pipes as building blocks to achieve reliable delivery, **with a caveat of having actual endpoints and pipe workers always running and not having errors**
+
+**A difference between pipe reliability and delivery reliability is that pipe does not take into account delivery between pipe workers and application workers**
+
+
+## Improving reliability
+
+Retries can be used to improve delivery reliability.
+
+Given a chance to deliver message from `A` to `B`: `R(A->B) = 0.5`
+
+If we retry delivery once, it becomes `R1 = R + (1- R) * R = 0.75`, `R2 = 0.875` etc.
+
+If retries `-> ∞`, then `R∞->1`
+
+Retries decrease uniqness rate
+Given `DM(A->B) = {0.5, 1}`
+`DMr(A->B)` - delivery mode of `A->B` with retries
+`DMr∞(A->B) = {->1, ->0.5}`
+
+To control retries, confirmation messages and timeouts are used
+
+<img src="./images/retries.jpg" width="100%">
+
+To provide at-least once delivery between workers which don't have additional logic,
```

Comment:
You may quote or use any other artifact to visually group the delivery types
```suggestion
To provide `at-least once` delivery between workers which don't have additional logic,
```

Reply:
For consistency will use at-least-once, at-most-once and exactly-once without quoting.